internal/adminapi/validators/consumer/credentials/validation.go

```diff
@@ -0,0 +1,110 @@
+// -----------------------------------------------------------------------------
+
+// Credential is a metadata struct to help validate the contents of
+// consumer credentials, particularly unique constraints on the underlying data.
+type Credential struct {
+	Key   string
+	Value []byte
+	Type  string
+}
+
+// IsKeyUniqueConstrained indicates whether or not a given key for a given credential
+// type has a unique constraint placed upon it in the Kong Admin API.
+func (c Credential) IsKeyUniqueConstrained() (constrained bool) {
+	constrainedKeys, credTypeHasConstraints := uniqueKeyConstraints[c.Type]
+	if !credTypeHasConstraints {
+		return
+	}
+
+	for _, constrainedKey := range constrainedKeys {
+		if c.Key == constrainedKey {
+			constrained = true
+			return
+		}
+	}
+
+	return
+}
+
+// -----------------------------------------------------------------------------
+// Validation - Credentials Lists
+// -----------------------------------------------------------------------------
+
+// Credentials is a list of Credential objects.
+type Credentials []Credential
+
+// DoesNewCredentialViolateUniqueConstraints indicates whether or not a unique constraint has been violated
+// given a list of credentials that have already been processed.
+func (cs Credentials) DoesNewCredentialViolateUniqueConstraints(newCredential Credential) bool {
+	// retrieve all the keys which are constrained for this type
+	constraints, ok := uniqueKeyConstraints[newCredential.Type]
+	if !ok {
+		return false // there are no constraints for this credType
+	}
+
+	// check each constrained key for this type
+	key := newCredential.Key
+	for _, constrainedKey := range constraints {
+		// if the key is constrainted, check the historical list for duplicate values
+		if key == constrainedKey {
+			for _, existing := range cs {
+				// if any duplicate values exist in the historical data, the unique constraint is violated
+				if bytes.Equal(existing.Value, newCredential.Value) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+// uniqueKeyConstraints a map of unique key constraints for any given credential type.
+// This map is the crux of all unique key constraint validation and is derived from
+// the relevant Lua code for the types in the backend Kong Admin API.
+//
+// Example: https://github.com/kong/kong/blob/master/kong/plugins/basic-auth/daos.lua
+//
+// So if you're in here doing maintenance and you need to add/remove constraints due
+// to upstream changes, check the "kong/plugins" directory of the upstream repo
+// for every given type.
+var uniqueKeyConstraints = map[string][]string{
+	"basic-auth": {"username"},
+	"hmac-auth":  {"username"},
+	"jwt":        {"key"},
+	"key-auth":   {"key"},
+	"oauth2":     {"client_id", "code", "access_token", "refresh_token"},
```

Comment:
@rainest please [double check me](https://github.com/Kong/kong/blob/master/kong/plugins/oauth2/daos.lua) on this one :thinking:

Reply:
just `client_id` from https://github.com/Kong/kong/blob/b0fe5611adc1cb34822dae460ed68be5582c9a33/kong/plugins/oauth2/daos.lua#L34

the other entities defined in that file aren't created from credential secrets. clients generate them dynamically via an API exposed on protected routes.

Reply:
sounds good, changed that in 78ac76536b909f12750063fdefb06a6aa2ef1187 :+1: